docs/ethernet.md

```diff
@@ -1,28 +1,118 @@
+and on Windows, and then describe the processing `vpnkit` performs on these
+frames in order to provide the illusion of true Internet connectivity.
 
-![ethernet diagram](http://docker.github.io/vpnkit/ethernet.png)
+## Plumbing inside Docker for Mac
 
-Frames arriving from the VM interface are processed by a simple ethernet switch.
-The switch demultiplexes traffic onto output ports by matching against rules.
-The current rules match only on the destination IPv4 address. Frames which
-don't match any rule are forwarded to a default port.
+The Docker for Mac VM is running on top of the [hyperkit](https://github.com/moby/hyperkit)
+hypervisor. The VM has a `virtio-vpnkit` PCI device which appears as a `virtio-net`
+network interface inside the VM. The
+[virtual hardware implementation](https://github.com/moby/hyperkit/blob/master/src/lib/pci_virtio_net_vpnkit.c) connects to
+`vpnkit` on the host over a Unix domain socket and encapsulates frames using a
+simple custom protocol.
+
+![Mac plumbing diagram](http://moby.github.io/vpnkit/mac.png)
+
+## Plumbing inside Docker for Windows
+
+The Docker for Windows VM is running on top of Hyper-V. `vpnkit` on the host
+uses Hyper-V sockets to connect to a process (`tap-vsockd`) inside the VM which accepts the
+connection and configures a `tap` device. Frames are encapsulated using the
+same custom protocol as on the Mac.
+
+![Windows plumbing diagram](http://moby.github.io/vpnkit/win.png)
+
+Note: the connection is currently made from the Host to the VM to work around
+a bug in older versions of Windows 10. At some point this should change to be
+a connection from the VM to the host.
+
+Note: the userspace `tap-vsockd` process in the VM which configures a `tap` device could be
+replaced with a custom Linux kernel driver which knows how to encapsulate the
+frames and communicate over Hyper-V sockets.
+
+## When a frame arrives in vpnkit
+
+Frames arriving from the VM are processed by a simple internal ethernet switch.
+The switch demultiplexes traffic onto output ports by matching on the destination
+IPv4 address. Frames which don't match any rule are forwarded to a default port.
 
 Frames arriving on the default port are examined and
 
 - if they contain ARP requests, we send a response using a static global ARP
-  tabl
-- if they contain IPv4 datagrams then we create a fresh TCP/IP stack, a fresh
-  switch port and connect them together so that all future IPv4 traffic to the
-  same destination address is processed by the new stack.
+  table
+- if they contain IPv4 datagrams then we create a fresh virtual TCP/IP endpoint
+  using the [Mirage](https://mirage.io/) [TCP/IP stack](https://github.com/mirage/mirage-tcpip)
+  (no kernel TCP/IP interfaces are involved), a fresh
+  switch port on our internal switch and connect them together so that all future IPv4 traffic to the
+  same destination address is processed by the new endpoint.
+
```

Comment:
There's also the L2 bridge in there now - so frames can come from any other Vpnkit instance via the bridge or from the connected VM: https://github.com/moby/vpnkit/blob/master/src/hostnet/slirp.ml#L731
 -- and frames from the VM without server/client/broadcast dst addresses will be sent to the bridge https://github.com/moby/vpnkit/blob/master/src/hostnet/slirp.ml#L749

The static ARP table also includes mac addresses to other Vpnkit instances.